week2/project/src/App.js

```diff
@@ -0,0 +1,79 @@
+
+import React, { useState, useEffect } from "react";
+import { BrowserRouter as Router } from "react-router-dom";
+import AppRoutes from "./AppRoutes";
+import "./App.css";
+import "./styles/ProductList.css";
+import "./styles/CategoryList.css";
+
+import { getCategories, getProducts } from "./services/api";
+
+const App = () => {
+  const [categories, setCategories] = useState([]);
+  const [selectedCategory, setSelectedCategory] = useState(null);
+  const [selectedProduct, setSelectedProduct] = useState(null);
+  const [loading, setLoading] = useState(true);
+  const [error, setError] = useState(null);
+  const [allProducts, setAllProducts] = useState([]);
+
+  useEffect(() => {
+    // Fetch categories and products data on component mount
+    const fetchData = async () => {
+      try {
+        const categoriesData = await getCategories();
+        const productsData = await getProducts();
+        setCategories(categoriesData);
+        setAllProducts(productsData);
+        setLoading(false);
+      } catch (error) {
+        setError("Error fetching data");
+        setLoading(false);
+      }
+    };
+
+    fetchData();
+  }, []);
+
+  const handleCategoryClick = (category) => {
+    // Handle category click
+    setSelectedCategory(category);
+    setSelectedProduct(null);
+  };
+
+  const renderLoadingOrError = () => {
+    // Render loading or error message
+    if (loading) {
+      return <p>Loading...</p>;
+    } else if (error) {
+      return <p>{error}</p>;
+    }
+  };
```

Comment:
Nice job adding the loading and error displaying in a separate place, not in `return` to make it more readable! 👍 

I would detach them even further to be honest, separate it even more and because they're not entirely related to each other. Also, I would not necessarily create a separate function for it because the final `return` should be responsible for rendering the main content.

```suggestion
    if (loading) {
      return <p>Loading...</p>
    }
    
    if (error) {
      return <p>{error}</p>
    }
```